stat/stat.go

```diff
@@ -988,24 +992,68 @@ func Quantile(p float64, c CumulantKind, x, weights []float64) float64 {
 	}
 	switch c {
 	case Empirical:
-		var cumsum float64
-		fidx := p * sumWeights
-		for i := range x {
-			if weights == nil {
-				cumsum++
-			} else {
-				cumsum += weights[i]
-			}
-			if cumsum >= fidx {
-				return x[i]
-			}
-		}
-		panic("impossible")
+		return empiricalQuantile(p, x, weights, sumWeights)
+	case AvgDiscont:
+		return avgDiscontQuantile(p, x, weights, sumWeights)
 	default:
 		panic("stat: bad cumulant kind")
 	}
 }
 
+func empiricalQuantile(p float64, x, weights []float64, sumWeights float64) float64 {
+	var cumsum float64
+	fidx := p * sumWeights
+	for i := range x {
+		if weights == nil {
+			cumsum++
+		} else {
+			cumsum += weights[i]
+		}
+		if cumsum >= fidx {
+			return x[i]
+		}
+	}
+	panic("impossible")
+}
+
+func avgDiscontQuantile(p float64, x, weights []float64, sumWeights float64) float64 {
+	// use algorithm from:
+	//  Hyndman, R.J and Fan, Y, (1996)
+	//  "Sample quantiles in statistical packages"
+	//  American Statistician, 50, 361-365
+	//
+	// also available from R's documentation:
+	//  https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html
+	fidx := p * sumWeights
```

Comment:
I can't find this algorithm in the paper you link (seems like it should be in section 2, but I don't see it). However, this really doesn't look like it works in general. If the weights are uneven, there's no reason that `p * sumWeights` should be close to the right place in the distribution. It may be a good initial guess, but this doesn't do any iteration to find the right answer. 

I think this should have a loop that looks like empiricalQuantile, but it returns `x[i] ` if `cumsum == fidx`, and returns `(x[i] + x[i+1])/2` if it's not equal.